File: app/src/main/java/com/cnrtflm/pixelpicker/LocaleHelper.java
package com.cnrtflm.pixelpicker;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;

import java.util.Locale;

public class LocaleHelper {

    private static final String PREFS_NAME = "app_settings";
    private static final String KEY_LANGUAGE = "language";

    /**
     * 获取保存的语言代码，如果没有保存则返回空字符串
     */
    public static String getSavedLanguage(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        return prefs.getString(KEY_LANGUAGE, "");
    }

    /**
     * 保存用户选择的语言
     */
    public static void saveLanguage(Context context, String languageCode) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        prefs.edit().putString(KEY_LANGUAGE, languageCode).apply();
    }

    /**
     * 应用语言设置，需要在 Activity.onCreate 之前调用
     */
    public static Context applyLanguage(Context context) {
        String language = getSavedLanguage(context);
        if (language.isEmpty()) {
            return context;
        }

        Locale locale = new Locale(language);
        Locale.setDefault(locale);

        Resources resources = context.getResources();
        Configuration config = resources.getConfiguration();
        config.setLocale(locale);
        config.setLayoutDirection(locale);

        return context.createConfigurationContext(config);
    }
}

File: app/src/main/java/com/pixelcolorpicker/util/LocaleHelper.kt
package com.pixelcolorpicker.util

import android.content.Context
import android.content.res.Configuration
import java.util.Locale

/**
 * 极简语言切换工具（不依赖 AppCompat）。
 *
 * 在 Activity.attachBaseContext 中调用 [wrap] 应用已保存的语言，
 * 切换语言时调用 [save] 并 recreate()。
 */
object LocaleHelper {

    private const val PREFS = "settings"
    private const val KEY_LANG = "app_language"

    /** 支持的语言代码。 */
    val SUPPORTED = listOf("zh", "en", "ja")

    fun getSavedLanguage(context: Context): String? {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val lang = prefs.getString(KEY_LANG, null)
        return if (lang != null && lang in SUPPORTED) lang else null
    }

    fun save(context: Context, lang: String) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putString(KEY_LANG, lang)
            .apply()
    }

    fun wrap(context: Context): Context {
        val lang = getSavedLanguage(context) ?: return context
        val locale = Locale(lang)
        Locale.setDefault(locale)
        val config = Configuration(context.resources.configuration)
        config.setLocale(locale)
        return context.createConfigurationContext(config)
    }
}